### PR TITLE
Relax `nogil_check()` of `SliceIndexNode` to support array initialisation

### LINKED ISSUE #1662: Cannot copy C arrays in nogil functions
This code:

```
cdef void foo() nogil:
    cdef double[16] x
    cdef double[16] y
    y[:] = x[:]
    y = x
```

throws the compile time error "Slicing Python object not allowed without gil" for both assignments. However, the generated code in both cases is a simple memcpy and doesn't require the GIL.

### LINKED ISSUE #3160: Initializing array with values requires GIL?
When trying to compile the following code with Cython, I get the error message below.

```cython
cdef void func() nogil: 
    cdef int[4] a = [0, 1, 2, 3]
```

```cython
/Users/jkirkham/miniconda/envs/cython/lib/python3.7/site-packages/Cython/Compiler/Main.py:369: FutureWarning: Cython directive 'language_level' not set, using 2 for now (Py2). This will change in a later release! File: /Users/jkirkham/test.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)

Error compiling Cython file:
------------------------------------------------------------
...
cdef void func() nogil: 
    cdef int[4] a = [0, 1, 2, 3]
        ^
------------------------------------------------------------

test.pyx:2:9: Slicing Python object not allowed without gil
```

Dropping the `nogil` makes the error go away. However the C code generated looks something like this and as such doesn't appear to be using the GIL.

```c
static void __pyx_f_4test_func(void) {
  ...

  int __pyx_t_1[4];

  __pyx_t_1[0] = 0;
  __pyx_t_1[1] = 1;
  __pyx_t_1[2] = 2;
  __pyx_t_1[3] = 3;

  ...
}
```

### PR DESCRIPTION
**Pull request by matusvalo**
Fixes #3160
Fixes #1662

### CODE REVIEW

**Comment by matusvalo** [comment]
I have just now found out discussion in #1663. Is there any way how to produce exception when two arrays has different length? Compiling following code:
```cython
    cdef double[16] x
    cdef double[17] y
    y[:] = x[:]
    y = x
```
produces compilation time error not runtime exception :-(

**Comment by h-vetinari** [comment]
> produces compilation time error not runtime exception :-(

How is that a bad thing? The arrays are known to be incompatible at compile-time, why shouldn't that generate a compilation error? Or do you mean the case if `16` / `17` were different variables here that the compiler cannot introspect?

**Comment by matusvalo** [comment]
> How is that a bad thing? 

I am not saying it is a bad think. I am trying to understand if raising an exception will affect GIL. In the original PR [1], this was reason for reverting similar PR as mine.

Hence I am trying to find out how to get exception when two arrays have different size.

[1] https://github.com/cython/cython/pull/1663#issuecomment-299685838

**Comment by da-woods** [comment]
> I am trying to understand if raising an exception will affect GIL

We typically allow exceptions to be raised in `nogil` blocks. This reacquires the GIL but this happens automatically within Cython. So I don't think a runtime exception is a problem.

**Comment by matusvalo** [comment]
Marking as ready for review. @da-woods if you have any other idea for additional tests, let me know and I will implement it to the tests.

**Comment by da-woods** [comment]
> Not sure if I understand correctly, but copying it requires GIL acquired

I was proposing something like

```
cdef f(double[:] a) nogil:
   cdef double[5] b = a[:]
```

I'm not sure without a very close look how closely related to your PR this is. But I think it shouldn't require the GIL, and gives you a variable slice size that Cython can't make fail at compile time.

Alternatively, something like:

```
cdef g(int N) nogil:
   cdef double[5] a = [1, 2, 3, 4, 5]
   cdef double[5] b = a[:N]
```

might give you a runtime failure, since it can't know `N` at compile time. Not sure if that works though.

These suggestions aren't based on a serious look though.

---------------------------------------------------

Will add this to the list of things I should look at in the fairly near future...

**Comment by matusvalo** [comment]
Thank you I will have a look on your proposals.

> Will add this to the list of things I should look at in the fairly near future...

REALLY appreciated. I understand that you have a huge pile of TODOs and not so much spare time. Let me know if I can somehow help you out.

**Comment by matusvalo** [comment]
I have checked your proposals and both of them fails due different errors

> **Note**: I checked both examples against master branch and with function allowing GIL. Still both examples fails with the same errors. Hence I conclude that the failures are not caused by this PR.

### 1st example

```cython
cdef double[4] a = [1.0, 2.0, 3.0, 4.0]

cdef f(double[:] a) nogil:
   cdef double[5] b = a[:]

f(a)
```


```
Error compiling Cython file:
------------------------------------------------------------
...


cdef double[5] a = [1.0, 2.0, 3.0, 4.0, 5.0]

cdef void f(double[:] a) nogil:
   cdef double[5] b = a[:]
                       ^
------------------------------------------------------------

test.pyx:8:23: Cannot assign type 'double[:]' to 'double [5]'

```

### 2nd example
Here fails slicing even with constant number.

```cython
cdef void g() nogil:
   cdef double[5] a = [1, 2, 3, 4, 5]
   cdef double[5] b = a[:6]

g()
```



```
Error compiling Cython file:
------------------------------------------------------------
...



cdef void g() nogil:
   cdef double[5] a = [1, 2, 3, 4, 5]
   cdef double[5] b = a[:6]
                       ^
------------------------------------------------------------

test.pyx:8:23: Slicing is not currently supported for 'double *'.
```

**Review by da-woods — COMMENTED** [review]
Right - I found the way to get an exception:

```
cdef void f(int n) nogil:
    cdef double[5] a = [1,2,3,4,5]
    cdef double[6] b
    b[:n] = a
```

it comes from `generate_slice_guard_code`. To me it doesn't look like it'll re-acquire the GIL before raising the exception.

The reasonable fixes would be:
* Re-strengthen the GIL check for this case
* Acquire the GIL around the exception-raising code (`code.put_ensure_gil` and `code.put_release_ensured_gil` I think)

Option 2 is the best, and I think should be straightforward

**Comment by matusvalo** [comment]
> Acquire the GIL around the exception-raising code (code.put_ensure_gil and code.put_release_ensured_gil I think)

Implemented.

File: tests/errors/nogil.pyx [hunk 1]
```diff
@@ -97,6 +97,15 @@ cdef int fstrings(int x, object obj) except -1 nogil:
...
     f"{obj}"
 
+cdef void slice_array() nogil:
+    with gil:
+        b = [1, 2, 3, 4]
+    cdef int[4] a = b[:]
+
+cdef int[:] main() nogil:
+    cdef int[4] a = [1,2,3,4]
+    return a
```
[the remark this answers is not in the document]
Comment (matusvalo) [reply tests/errors/nogil.pyx:107]: > doesn't matter for the sake of an error test

Yes I did not care about dangling pointers. But if you want I can change the test.

File: tests/run/nogil.pyx [hunk 2]
```diff
@@ -154,3 +159,12 @@ def test_copy_array3():
...
     return copy_array3()
+
+def test_copy_array_exception(n):
+    """
+    >>> test_copy_array_exception(20)
+    Traceback (most recent call last):
+        ...
+    ValueError: Assignment to slice of wrong length, expected 5, got 20
+    """
+    return copy_array_exception(n)
```
Comment (da-woods) [inline tests/run/nogil.pyx:170]:
```suggestion
    copy_array_exception(n)
    
def test_copy_array_exception_nogil(n): 
    >>> test_copy_array_exception_nogil(20)
    Traceback (most recent call last):
        ...
    ValueError: Assignment to slice of wrong length, expected 5, got 20
    """
    with nogil:
        copy_array_exception(n)
```

**Comment by matusvalo** [comment]
CI should be fixed now

**Comment by da-woods** [comment]
Thanks. I think that probably covers the issues with the old PR for this

**Pull request merged**